File: android/app/src/main/java/com/sovereignnetworkmobile/web4/Web4ManifestCache.kt
package com.sovereignnetworkmobile.web4

import java.util.LinkedHashMap

class Web4ManifestCache(private val maxEntries: Int = 64) {
    private val cache: LinkedHashMap<String, Pair<String, Web4Manifest>> =
        object : LinkedHashMap<String, Pair<String, Web4Manifest>>(16, 0.75f, true) {
            override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, Pair<String, Web4Manifest>>?): Boolean {
                return size > maxEntries
            }
        }

    @Synchronized
    fun put(domain: String, manifestCid: String, manifest: Web4Manifest) {
        cache[domain] = manifestCid to manifest
    }

    @Synchronized
    fun get(domain: String): Pair<String, Web4Manifest>? {
        return cache[domain]
    }
}

File: ios/Web4ManifestCache.swift
import Foundation

final class Web4ManifestCache {
  private var cache: [String: (manifestCid: String, manifest: Web4Manifest)] = [:]
  private let lock = NSLock()

  func get(domain: String) -> (String, Web4Manifest)? {
    lock.lock(); defer { lock.unlock() }
    return cache[domain]
  }

  func put(domain: String, manifestCid: String, manifest: Web4Manifest) {
    lock.lock(); defer { lock.unlock() }
    cache[domain] = (manifestCid, manifest)
  }
}
